// openclient/nova/v1.1/services.js
var async = require('async'),
    base = require("../../client/base"),
    error = require("../../client/error"),
    urljoin = require("../../client/utils").urljoin;


var ServiceManager = base.Manager.extend({
  namespace: "os-services",
  plural: "services",
  singular: "service",

  action: function (params, callback) {
    var url = urljoin(this.get_base_url(params), params.id || params.data.id, "action");
    params = this.prepare_params(params, url, "singular");
    params.data = params.data.cluster;
    return this.client.post(params, callback);
  },

  enable: function (params, callback) {
    params.user_raw_data = true;
    var url = this.get_base_url(params) + "/enable";
    params = this.prepare_params(params, url, "singular");
    params.data = params.data.service;
    return this.client.put(params, callback);
  },

  disable: function (params, callback) {
    params.user_raw_data = true;
    var url = this.get_base_url(params) + "/disable";
    params = this.prepare_params(params, url, "singular");
    params.data = params.data.service;
    return this.client.put(params, callback);
  }
});


module.exports = ServiceManager;
