// openclient/workflow/v1.0/client.js
var base = require("../../client/base");

var WorkflowRequestTypeManager = require("./workflow_request_type"),
    FeedbackManager = require("./feedback"),
    OptionManager = require("./option"),
    GroupManager = require("./group"),
    RegisterManager = require("./register"),
    WorkflowRequestManager = require("./workflow_request"),
    AppManager = require("./app"),
    TemplateManager = require("./load_template"),
    FeedbackReplyManager = require("./feedback_reply");


var Workflow = base.Client.extend({
  service_type: "workflow",
  version: "1.0",

  init: function (options) {
    this._super(options);

    this['workflow-request-types'] = new WorkflowRequestTypeManager(this);
    this['apps'] = new AppManager(this);
    this['workflow-requests'] = new WorkflowRequestManager(this);
    this.feedbacks = new FeedbackManager(this);
    this.options = new OptionManager(this);
    this['option_groups'] = new GroupManager(this);
    this.register = new RegisterManager(this);
    this['feedback_replies'] = new FeedbackReplyManager(this);
    this['load_template'] = new TemplateManager(this);
  }
});

module.exports = Workflow;
