// openclient/workflow/v1.0/workflow_request.js
var base = require("../../client/base"),
    error = require("../../client/error"),
    urljoin = require("../../client/utils").urljoin;


var WorkflowRequestManager = base.Manager.extend({
    namespace: "workflow-requests",

    all: function (params, callback) {
        params.result_key = "workflow-request";
        params.use_raw_data = true;
        return this._super(params, callback);
    },

    create: function (params, callback) {
        params.url = this.get_base_url(params);
        this.client.post(params, callback);
    },

    update: function (params, callback) {
        params.url = urljoin(this.get_base_url(params), params.id);
        this.client.put(params, callback);
    },

    edit: function (params, callback) {
        params.url = urljoin(this.get_base_url(params), params.id);
        params.url = params.url + "/edit";
        this.client.get(params, callback);
    },

    resourceCheck: function (params, callback) {
        params.url = urljoin(this.get_base_url(params));
        delete params.result_key;
        params.url = params.url.replace('workflow-requests', 'resource_check');
        this.client.post(params, callback);
    }
});


module.exports = WorkflowRequestManager;
