// openclient/cinder/v2/quotas.js
var base = require("../../client/base"),
    error = require("../../client/error");
    urljoin = require("../../client/utils").urljoin;
// TODO(roland): Make methods throw NotImplemented where applicable.

var QuotaManager = base.Manager.extend({
  namespace: "os-quota-sets",
  singular: 'quota_set',
  plural: 'quota_sets',

  _quota_names: ["volumes", "gigabytes"],

  update: function (params, callback) {
    params.id = params.id || params.data.id;

    // Treat blank values as "unlimited" and set them to -1.
    this._quota_names.forEach(function (name) {
      var val = params.data[name];
      if (typeof val !== "undefined" && val !== 0 && !val) params.data[name] = -1;
    });

    params.parseResult = function (result) {
      result.id = params.id;
      return result;
    };

    this._super(params, callback);
  },
  defaults: function (params,callback) {
    params.parseResult = function (result) {
      result.disk = -1;
      result["OS-FLV-EXT-DATA:ephemeral"] = -1;
      return result;
    };
    params.manager_method = "get";
    params = this.normalize_id(params);
    var url = urljoin(this.get_base_url(params),params.id,"defaults");
    params = this.prepare_params(params,url,"singular");
    return this.client[params.http_methon || this.method_map.get](params,callback);
  },

  usages: function (params, callback) {
    var manager = this,
        usages = {},
        flavors = {};

    usages.id = this.client.tenant.id;
    usages.gigabytes = 0;
    usages.disk = 0;

    this.client.volumes.all({detail: true}, function (err, volumes) {
      if (err) return manager.safe_complete(err, null, null, params, callback);

      usages.volumes = volumes.length;
      volumes.forEach(function (volume) {
        usages.gigabytes += volume.size;
        usages.disk += volume.size;
      });

      manager.safe_complete(err, usages, {status: 200}, params, callback);
    });
  }
});


module.exports = QuotaManager;
