// openclient/neutron/v2.0/client.js
var base = require("../../client/base"),
    urljoin = require("../../client/utils").urljoin;

var NetworkManager = require('./networks'),
    SubnetManager = require('./subnets'),
    PortManager = require('./ports'),
    RouterManager = require('./routers'),
    FloatingIPManager = require('./floatingips'),
    SecurityGroupManager = require('./security-groups'),
    SecurityGroupRulesManager = require('./security-group-rules');
    LbaasManagers = require('./lbaas');

var Neutron = base.Client.extend({
    service_type: "network",   

    version: "2.0",

    init: function (options) {
      if (options.url) {
          options.url = urljoin(options.url, 'v2.0');
      }
      this._super(options);
      this.networks = new NetworkManager(this);
      this.subnets = new SubnetManager(this);
      this.ports = new PortManager(this);
      this.routers = new RouterManager(this);
      this.floatingips = new FloatingIPManager(this);
      this['security-groups'] = new SecurityGroupManager(this);
      this['security-group-rules'] = new SecurityGroupRulesManager(this);
      var lbaasComs = LbaasManagers();
      for (var com in lbaasComs) {
          var comManager = new lbaasComs[com](this);
          var key = comManager['plural'];
          this[key] = comManager;
      }
    }
});

module.exports = Neutron;
