// openclient/nova/v2.1/client.js
var base = require("../../client/base");

var AZManager = require("./availability_zones"),
    CertsManager = require("./certs"),
    FlavorManager = require("./flavors"),
    FloatingIPManager = require('./floating_ips'),
    FloatingIPPoolManager = require('./floating_ip_pools'),
    FloatingIPBulkManager = require('./floating_ip_bulk'),
    ImageManager = require('./images'),
    KeypairManager = require('./keypairs'),
    LimitManager = require('./limits'),
    NetworkManager = require('./networks'),
    QuotaManager = require('./quotas'),
    QuotaClassManager = require('./quota_class'),
    SecurityGroupManager = require('./security_groups'),
    SecurityGroupRuleManager = require('./security_group_rules'),
    SecurityGroupDefaultRuleManager = require('./security_group_default_rules'),
    ClusterManager = require('./clusters'),
    ServiceManager = require('./services'),
    HypervisorManager = require('./hypervisors'),
    ServerManager = require("./servers"),
    UsageManager = require("./usage");


var Nova = base.Client.extend({
  service_type: "compute",
  version: "2.1",

  init: function (options) {
    this._super(options);

    this.availability_zones = new AZManager(this);
    this.certs = new CertsManager(this);
    this.flavors = new FlavorManager(this);
    this.floating_ips = new FloatingIPManager(this);
    this.floating_ip_pools = new FloatingIPPoolManager(this);
    this.floating_ip_info = new FloatingIPBulkManager(this);
    this.images = new ImageManager(this);
    this.keypairs = new KeypairManager(this);
    this.limits = new LimitManager(this);
    this["os-networks"] = new NetworkManager(this);
    this.quotas = new QuotaManager(this);
    this.quota_class = new QuotaClassManager(this);
    this.security_groups = new SecurityGroupManager(this);
    this.security_group_rules = new SecurityGroupRuleManager(this);
    this.security_group_default_rules = new SecurityGroupDefaultRuleManager(this);
    this.servers = new ServerManager(this);
    this.clusters = new ClusterManager(this);
    this.services = new ServiceManager(this);
    this.hypervisors = new HypervisorManager(this);
    this.usage = new UsageManager(this);
  }
});

module.exports = Nova;
