// openclient/index.js
var base = require("./client/base"),
    error = require("./client/error");

var openclient = {
  // Base functional components.
  Client: base.Client,
  Manager: base.Manager,
  error: error,
  api_map: {
    'openstack': {
      'identity': require('./keystone/versions'),
      'compute': require('./nova/versions'),
      'image': require('./glance/versions'),
      'volume': require('./cinder/versions'),
      'network': require('./neutron/versions'),
      'object-store': require('./swift/versions'),
      'orchestration': require('./heat/versions'),
      'metering': require('./ceilometer/versions'),
      'workflow': require('./workflow/versions'),
      'deployment': require('./deployment/versions'),
      'alarm': require('./alarm/versions'),
      'maintenance': require('./highland/versions'),
      'price': require('./funnel/versions'),
      'pubcloud': require('./duckbill/versions'),
      'recovery': require('./egis/version'),
    }
  },

  getAPI: function (provider, type, version) {
    var api = openclient.api_map[provider][type];
    if (version) {
      return api[version.toString()];
    } else {
      // TODO: Add the ability for this function to do version discovery.
      return api[api.current];
    }
  }
};


// Left for backwards compatibility. DEPRECATED.
// Hard-coded OpenStack-specific clients by codename.
[
  ['Cinder', require('./cinder/v1/client')],
  ['Glance', require('./glance/v1.0/client')],
  ['Keystone', require("./keystone/v2.0/client")],
  ['Nova', require("./nova/v1.1/client")],
  ['Neutron', require("./neutron/v2.0/client")],
  ['Swift', require("./swift/v1/client")],
  ['Heat', require("./heat/v1/client")],
  ['Deployment', require("./deployment/v1.0/client")],
  ['Highland', require("./highland/v1.0/client")],
  ['Funnel', require("./funnel/v1.0/client")],
  ['Duckbill', require("./duckbill/v1/client")],
  ['Egis',require("./egis/v1/client")],
].forEach(function (pair) {
  Object.defineProperty(openclient, pair[0], {
    get: function () {
      console.trace("Use of openclient." + pair[0] + " is deprecated. Please use the openclient.getAPI method instead.");
      return pair[1];
    },
    enumerable: true,
    configurable: true
  });
});


module.exports = openclient;
